### PR TITLE
Add SSRF protection with centralized HTTP request wrapper

### LINKED ISSUE #619: Validate URLs in metadata and callback evaluators to prevent malformed URLs
## Problem

When using templated URLs (JSON placeholders or CEL expressions) in evaluators that make HTTP calls, user-controlled values can result in malformed URLs being constructed. This can lead to unexpected behavior or security issues.

Affected evaluators:
- Generic HTTP metadata (`metadata.http`)
- OIDC UserInfo metadata
- UMA resource metadata
- OAuth2 token introspection
- HTTP callbacks

## Proposed Solution

1. **Add URL validation** after template/CEL resolution:
   - Parse resolved URLs with `url.Parse()`
   - Reject non-HTTP(S) schemes
   - Return clear error messages for invalid URLs

2. **Document SSRF risks** in CRD field documentation:
   - Add warnings in `api/v1beta3/auth_config_types.go` about risks of using user-controlled values in URL templates
   - Educate users to avoid using untrusted input (e.g., JWT claims, headers) directly in URL construction

## Acceptance Criteria

- [ ] All evaluators making HTTP calls validate URLs after template resolution
- [ ] Invalid URLs result in clear error messages
- [ ] CRD godoc comments warn about SSRF risks when using dynamic URL construction
- [ ] Tests cover malformed URL scenarios

### PR DESCRIPTION
**Pull request by guicassolato**
Closes #619

## Summary

This PR addresses SSRF (Server-Side Request Forgery) security concerns by adding URL validation to all outbound HTTP requests and refactoring HTTP request handling into a centralized package with cleaner dependency architecture.

### URL Validation
- **New `ValidateURL()` function** performs comprehensive checks on URLs before making HTTP requests:
  - ✅ Enforces `http` or `https` schemes only (blocks `file://`, `ftp://`, `javascript:`, etc.)
  - ✅ Validates host is present
  - ✅ Detects control characters (newlines, tabs, null bytes) that could enable request smuggling
  - ✅ Enforces maximum URL length (2048 characters) to prevent DoS
  - ✅ Validates URL parses correctly
  - ⚠️ **Does NOT block private IPs** - internal metadata and service calls are legitimate use cases

### Documentation
- Added **SSRF risk warnings** to 5 CRD fields that accept user-controlled URLs:
  - `HttpEndpointSpec.Url` and `UrlExpression`
  - `UserInfoMetadataSpec.UserInfoUrl`
  - `UmaMetadataSpec.Endpoint`
  - `OAuth2TokenIntrospectionSpec.Url`
  - `OAuth2ClientAuthentication.TokenUrl`

### Updated HTTP Calls
All evaluators now use the new validated request wrapper:
- `pkg/evaluators/metadata/generic_http.go`
- `pkg/evaluators/metadata/user_info.go`
- `pkg/evaluators/metadata/uma.go`
- `pkg/evaluators/identity/oauth2.go`
- `pkg/evaluators/authorization/opa.go`

### New `pkg/http` Package
Created centralized HTTP utilities with clean interfaces:
- `NewRequest(ctx, method, url, body)` - validates URL before creating request
- `NewRequestWithCredentials(ctx, method, url, body, credentialInjector, credentialValue)` - injects credentials into requests
- `NewClient()` - placeholder for future timeout configuration (#620)
- `CredentialLocation` interface for credential injection

### Dependency Inversion
- **Before**: `pkg/auth` was a leaf dependency
- **After**: `pkg/http` has no dependencies; `pkg/auth.AuthCredentials` embeds `http.CredentialLocation`
- Eliminates circular dependency concerns and improves modularity

### Improved Naming
Renamed methods for better clarity:
- `GetCredentialsIn()` → `GetPlacement()` - describes *where* credential goes
- `GetCredentialsKeySelector()` → `GetIdentifier()` - describes the key/name/prefix
- `GetCredentialsFromReq()` → `GetCredentialsFromAuthReq()` - clarifies it extracts from Envoy auth requests

### Consolidated Constants
- Moved credential location constants to `pkg/http`:
  - `InAuthorizationHeader`
  - `InCustomHeader`
  - `InCookie`
  - `InQuery`
- All code now uses these constants instead of string literals
- Updated `controllers/auth_config_controller.go` to use constants

### API Improvements
- `NewAuthCredential(placement, identifier)` - placement first as primary characteristic
- Consistent parameter ordering across related functions

## Testing

- ✅ 100% test coverage on `pkg/http`
- ✅ All existing tests pass
- ✅ Updated all tests to use constants instead of string literals
- ✅ 47 test cases in `pkg/http/request_test.go` covering validation edge cases
- ✅ Regenerated mocks for both `pkg/http` and `pkg/auth`

## Breaking Changes

⚠️ **Interface changes** - this is a breaking change for anyone implementing `AuthCredentials`:
- Method renames: `GetCredentialsIn()` → `GetPlacement()`, `GetCredentialsKeySelector()` → `GetIdentifier()`, `GetCredentialsFromReq()` → `GetCredentialsFromAuthReq()`
- `NewAuthCredential` parameter order changed from `(identifier, placement)` to `(placement, identifier)`

## Future Work

This PR addresses items 1 and 2 from #619. Item 3 (configurable HTTP client timeouts) will be addressed in a separate PR (#620).

## Summary by CodeRabbit

* **Bug Fixes**
  * Added prominent SSRF warnings across AuthConfig CRD docs and manifests for externally‑influenced URL fields (OAuth2, OIDC, HTTP metadata, UMA).

* **Refactor**
  * Centralised outbound HTTP handling and credential placement into safer helpers with URL validation, standardized request creation and context-aware metadata/UMA fetches.

* **Tests**
  * Expanded unit tests for URL validation, request creation with credentials, and related metadata/UMA flows.

### CODE REVIEW

File: pkg/http/request.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+		return fmt.Errorf("invalid URL: missing host")
+	}
+
+	return nil
+}
+
+// NewClient creates an HTTP client for making requests.
+// Currently returns http.DefaultClient, but will support configurable timeouts in the future (issue #620).
+func NewClient() *http.Client {
+	return http.DefaultClient
```
[the remark this answers is not in the document]
Comment (guicassolato) [reply pkg/http/request.go:161]: It will be addressed in #620

File: pkg/service/auth_pipeline_test.go [hunk 2]
```diff
@@ -564,8 +564,8 @@ func BenchmarkAuthPipeline(b *testing.B) {
...
...yI6eyJyZWFsbS1tYW5hZ2VtZW50Ijp7InJvbGVzIjpbInZpZXctaWRlbnRpdHktcHJvdmlkZXJzIiwidmlldy1yZWFsbSIsIm1hbmFnZS1pZGVudGl0eS1wcm92aWRlcnMiLCJpbXBlcnNvbmF0aW9uIiwicmVhbG0tYWRtaW4iLCJjcmVhdGUtY2xpZW50IiwibWFuYWdlLXVzZXJzIiwicXVlcnktcmVhbG1zIiwidmlldy1hdXRob3JpemF0aW9uIiwicXVlcnktY2xpZW50cyIsInF1ZXJ5LXVzZXJzIiwibWFuYWdlLWV2ZW50cyIsIm1hbmFnZS1yZWFsbSIsInZpZXctZXZlbnRzIiwidmlldy11c2VycyIsInZpZXctY2xpZW50cyIsIm1hbmFnZS1hdXRob3JpemF0aW9uIiwibWFuYWdlLWNsaWVudHMiLCJxdWVyeS1ncm91cHMiXX0sImFjY291bnQiOnsicm9sZXMiOlsibWFuYWdlLWFjY291bnQiLCJtYW5hZ2UtYWNjb3VudC1saW5rcyJdfX0sInNjb3BlIjoicHJvZmlsZSBlbWFpbCIsInNpZCI6IjExN2QxNzU2LWYzZGUtNGMzYy05YTBlLTNiNTkzNzZiYjRlOCIsImVtYWlsX3ZlcmlmaWVkIjpmYWxzZSwibmFtZSI6IlBldGVyIFdobyIsInByZWZlcnJlZF91c2VybmFtZSI6InBldGVyIiwiZ2l2ZW5fbmFtZSI6IlBldGVyIiwiZmFtaWx5X25hbWUiOiJXaG8iLCJlbWFpbCI6InBldGVyQGt1YWRyYW50LmlvIn0.Yy2aWR6_u0NBLx8x--OToYipfQ1f1KcC8zedsKDiymcbBiAaxrBQmaV2JC1PQVEgyxwmyMk0Rao2MdKGWk6pXB9mTUF5FX-pS8mkPIMUt1UVGJgzq7WR9KfRqdZSzRtFQHoDmTeA1-msayMYTAD8xtUH4JYRNbIXjY2cEtn8LjuLpQVR3DR4_ARMrEYXiDBS3rmmFKHdipqU7ozwJ_gtpZv8vfeiO3mUPyQLJKQ-nKpe_Z5z7tm_Ewh5MN2oBfn_0pcdANB3pe2RclGAm-YHlyNDTnAZL2Y1gdCmwzwigk7AJcgWtPqnRzvEQ9zRBxQRai5W5aNKYTxuKIG8k9N05w", nil).MinTimes(1)
```
[the remark this answers is not in the document]
Comment (guicassolato) [reply pkg/service/auth_pipeline_test.go:568]: Out of scope of this PR

File: pkg/evaluators/identity/plain.go [hunk 3]
```diff
@@ -28,18 +26,14 @@ func (p *Plain) Call(pipeline auth.AuthPipeline, ctx context.Context) (interface
...
 }
 
-func (p *Plain) GetCredentialsKeySelector() string {
+func (p *Plain) GetIdentifier() string {
 	return p.Pattern
 }
 
-func (p *Plain) GetCredentialsIn() string {
+func (p *Plain) GetPlacement() string {
 	return p.Pattern
```
[the remark this answers is not in the document]
Comment (guicassolato) [reply pkg/evaluators/identity/plain.go]: `Plain` only implements `auth.AuthCredentials` as a requirement to qualify as an identity evaluator. I've changed so only `GetIdentifier()` returns `Pattern`. This is the closest to a credential location associated with this struct. `GetPlacement()` now returns an empty string. Either way, neither of these functions implemented by `Plain` is called anywhere in the code.

File: pkg/auth/credentials_test.go [hunk 4]
```diff
@@ -206,27 +203,79 @@ func TestGetCredentialsFromQueryFail(t *testing.T) {
...
-	assert.Equal(t, req.Header.Get("Authorization"), creds.KeySelector+" 123")
-}
-
-func TestBuildRequestWithCredentialsEmpty(t *testing.T) {
-	creds := NewAuthCredential("", "")
-	req, err := creds.BuildRequestWithCredentials(context.TODO(), "http://example.com", "GET", "", nil)
+func TestGetCredentialsFromQueryWithSpecialCharactersInIdentifier(t *testing.T) {
+	// Test that special regex characters in the identifier don't cause panics
+	// This would have panicked with the old regexp.MustCompile approach
+	testCases := []struct {
```
Comment (maleck13) [inline pkg/auth/credentials_test.go:217]: would it be valuable to test malformed paths and add tests for some err cases?

File: pkg/evaluators/metadata/generic_http_test.go [hunk 5]
```diff
@@ -235,8 +230,7 @@ func TestGenericHttpCallWithCustomHeaders(t *testing.T) {
 	_, err := metadata.Call(pipelineMock, ctx)
 
 	assert.NilError(t, err)
-	assert.Equal(t, httpRequestMock.Header.Get("X-Requested-By"), "authorino")
-	assert.Equal(t, httpRequestMock.Header.Get("Content-Type"), "text/plain")
+	// Note: Can't check headers on httpRequestMock since it's no longer created in the test
```
Comment (maleck13) [inline pkg/evaluators/metadata/generic_http_test.go]: Do you intend to look at an improvement here? To be clear I am not sure. But is it a possible place for a regression to sneak in?

**Review by maleck13 — DISMISSED** [review]
LGTM some small suggestions 

- add some negative / error case tests if there is value
- look at introducing some form of header validation test back
- def want to address the timeout issue with the default http client (which you already called out)

File: pkg/evaluators/metadata/generic_http_test.go (hunk 5 above)
Comment (guicassolato) [reply pkg/evaluators/metadata/generic_http_test.go]: Right. After the refactoring, this test became limited to testing setting a custom header won't return an error.

Fixed it by using go native http test server and using a closure to capture the received headers.

Now, we should help prevent any regression.

Thanks, @maleck13 !

**Review by maleck13 — APPROVED** [review]
👍
